README.md
## Achievement Tracker
 - Track achievement progress on screen 
 - Track primary achievement in dedicated window

### Description:
Track all the achievements!

### Options:
    Show Failure Messages
        On / Off
    
    Show Primary Achievement Window
        On / Off

    Primary Achievement: Select which achievement shows in Primary Achievement Window
        [All The Achievements]

    Message Length: Time Displayed on Screen
        1 - 60 seconds

    Message Delay (Sols): Delay between messages in sols
        0 - 7 sols

    Message Delay (Hours): Delay between messages in hours (combines with sols)
        0 - 25 hours

### Special Thanks:
Thanks to *ChoGGi* for all of his assistance, ECM, and countless mods providing good example code for me to learn.

Thanks to *LoneSamurai* for the mod idea, testing, and suggestions.

Thanks to *Ella* for testing, feedback, and suggestions. 



## Notes:
If you have any feedback please contact me directly instead of reporting the mod. I'm always happy to take requests as well.<br>
I hang out in [*Discord*](https://discord.gg/kTgYq9UjWx) and on [*Reddit*](https://www.reddit.com/user/Fizzle_Fuze)

Please take a second to rate this mod and leave a comment.<br>
Your feedback helps me produce better quality mods!

Have a good game,<br>
Fizzle Fuze :)

## Info:

### Changelog:
Details on the latest changes as well as changes in future versions are found in [*Info/changelog.md*](Info/changelog.md)

### Contact Me:
I appreciate feedback on my mods and ideas for new ones!
 - Discord:https://discord.gg/kTgYq9UjWx
 - Reddit: /u/Fizzle_Fuze (/r/SurvivingMars)
 - Email: mods@fizzlefuze.com

See [*Info/ContactMe.txt*](Info/ContactMe.txt) for more info on contacting me.

### Other Locations:
My mods are also available on Steam and the Paradox Website:

[Steam Workshop](https://steamcommunity.com/id/fizzle_fuze/myworkshopfiles/?appid=464920) <br>
[Paradox Plaza](https://mods.paradoxplaza.com/games/surviving_mars?search=Fizzle_Fuze&sortBy=best)


### License:
See [*Info/License.md*](Info/license.md) and [*Info/LICENSE*](Info/LICENSE) for copyright and licensing info.


changelog.md
Version 1.7
 - Added option for length of time message stays on screen
 - Implemented work around for "real-time" changing based on game speed
 - Fixed scale for resource based achievements again XD
 - Forced re-initialization of objects so that they display correctly when loading a game saved with an older version
 - Improved object initialization to avoid hangs
 - Removed delay in updating PA text
 - Improved PA window
   - Added styling
   - Added close on double click
   - Added mod option to show PA or not

Version 1.6
- Fixed timer for messages
- Improved primary achievement display

Version 1.5
 - Show "proper" names in mod option for Primary Achievement
 - Added display showing primary achievement progress on the top right of the screen
 - Clicking on messages now shows further details on how to obtain the achievement
 - Added optional achievement failed message
 - Only tracks achievements for available DLCs
 - Added "Achievement Complete" message
 - Fixed the scale for the following achievements
   - IndiaConvertedWasteRock
   - BrazilConvertedWasteRock
   - RussiaExtractedAlot
   - BlueSunProducedFunding
   - BlueSunExportedAlot
   
Version 1.4
- Formatted numbers in messages
- Replaced DataInstances.Achievement with AchievementPresets
- Added options for frequency of notifications
- Duplicate messages no longer show
- Added targets for the following achievements
   - AsteroidHopping 
   - USAResearchedEngineering
   - Multitasking 
   - SpaceDwarves 
   - Willtheyhold 
   - SpaceExplorer
   - ScannedAllSectors
   - DeepScannedAllSectors

Version 1.3
- Fixed a bug with rocket landing count
- Consolidated rocket landing functions
- Completed achievements no longer show

Version 1.2
- Fixed the city the objects were created in. Thanks ChoGGi :)

Version 1.1
 - Initial creation

Code/OSD.lua
-- See license.md for copyright info
local function Log(...)
    FF.Funcs.LogMessage(CurrentModDef.title, "OSD", ...)
end

--locals
local UpdateThread

function CreateOSD()
    local ShowMAWindow = SharedModEnv["FF_AT_ShowMainAchievementWindow"]

    local function OnClick(...)
        --2do: change achieve
    end

    local function OnDoubleClick(...)
        ShowMAWindow = false
        Dialogs.InGameInterface.FF_AT_PrimaryAchievement:Close()
    end

    local Parent = Dialogs.InGameInterface
    if not Parent then
        Log("ERROR", "No InGameInterface!")
        return
    end

    if Parent.FFAT_PrimaryAchievement then --backwards compatibility
        Parent.FFAT_PrimaryAchievement:DeleteChildren()
        Parent.FFAT_PrimaryAchievement:delete()
    end

    if Parent.FF_AT_PrimaryAchievement then
        Parent.FF_AT_PrimaryAchievement:DeleteChildren()
        Parent.FF_AT_PrimaryAchievement:delete()
    end

    if not ShowMAWindow then
        return
    end

    local OSD = XWindow:new({
        ChildrenHandleMouse = false,
        Dock = "box",
        FoldWhenHidden = true, --how to hide?
        HAlign = "center",
        HandleKeyboard = false,
        HandleMouse = true,
        Id = "FF_AT_PrimaryAchievement",
        IdNode = true,
        Margins = box(0, 85, 0, 0), --below InfoBar
        OnMouseButtonDoubleClick = OnDoubleClick,
        OnMouseButtonDown = OnClick,
        Padding = box(2,2,2,2),
        RolloverHint = FF.Funcs.Translate("Keep up the good work! =)"),
        RolloverTemplate = "Rollover",
        RolloverText = FF.Funcs.Translate("<image UI/Infopanel/left_click.tga 1400>*2 Close"),
        RolloverTitle = FF.Funcs.Translate("Primary Achievement"),
        transparency = 50,
        VAlign = "top",
    }, Parent)

    local Frame = XFrame:new({
        Dock = "box",
        FrameBox = box(0,20,0,0),
        HAlign = "stretch",
        HandleKeyboard = false,
        Id = "FF_AT_PA_Frame",
        Image = "UI/CommonNew/ip.dds",
        Padding = box(5,0,5,5),
        VAlign = "stretch",
    }, OSD)

    XText:new({
        Dock = "top",
        HandleKeyboard = false,
        Id = "FF_AT_PA_Title",
        TextStyle = "FF_PrimaryAchievement_Title",
        TextHAlign = "center",
    }, Frame)

    XText:new({
        Dock = "top",
        HandleKeyboard = "false",
        Id = "FF_AT_PA_Description",
        Margins = box(5,10,5,0),
        TextStyle = "FF_PrimaryAchievement_Text",
    }, Frame)

    XText:new({
        Dock = "top",
        HandleKeyboard = "false",
        Id = "FF_AT_PA_Progress",
        TextStyle = "FF_PrimaryAchievement_Progress",
        TextHAlign = "center",
        VAlign = "stretch",
    }, Frame)

    OSD:SetParent(Parent) --2do: test if this is necessary
    UpdateOSD()
end

function UpdateOSD()

    local function UpdateText()
        local Title, Description, Progress

        for _, A in pairs(MainCity.labels.TrackedAchievement) do
            if A.Name == CurrentModOptions:GetProperty("ShowOnScreen") then

                local OSD = Dialogs.InGameInterface.FF_AT_PrimaryAchievement.FF_AT_PA_Frame
                if not OSD then
                    Log("No OSD")
                    return
                end

                Title = A.Name
                Description = A.Description
                Progress = "Current Progress: " .. FF.Funcs.FormatNumber(A.Value) .. " / " .. FF.Funcs.FormatNumber(A.Target)

                if not OSD.FF_AT_PA_Title or not OSD.FF_AT_PA_Description or not OSD.FF_AT_PA_Progress then
                    Log("ERROR", "Primary Achievement children don't exist!")
                    return
                end
                OSD.FF_AT_PA_Title:SetText(FF.Funcs.Translate(Title))
                OSD.FF_AT_PA_Description:SetText(FF.Funcs.Translate(Description))
                OSD.FF_AT_PA_Progress:SetText(FF.Funcs.Translate(Progress))
                UpdateThread = nil
            end
        end
    end

    if not UpdateThread then
            UpdateThread = CreateGameTimeThread(UpdateText)
    end
end

metadata.lua
return PlaceObj('ModDef', {
	'title', "Achievement Tracker",
	'description', "This mod is still a work in progress. Please hop in discord to help test and provide feedback. :)   https://discord.gg/kTgYq9UjWx",
	'image', "Images/thumbnail.png",
	'last_changes', "https://github.com/Surviving-Mars-Mods/Achievement-Tracker",
	'ignore_files', {
		"*.git/*",
		"*.svn/*",
		"src/*",
		"Info/2do.txt",
		"*.iml",
		".idea/*",
		"*.png",
	},
	'dependencies', {
		PlaceObj('ModDependency', {
			'id', "FIZZLE1",
			'title', "Fizzle Fuze's Library",
			'version_major', 1,
			'version_minor', 5,
		}),
	},
	'id', "FIZZLE3",
	'steam_id', "2755868001",
	'pops_desktop_uuid', "52a6e8c9-c772-45c5-a678-8db6896a8bcb",
	'pops_any_uuid', "3dab9f78-3c6f-4c5b-8ee7-9dfba90b9512",
	'author', "Fizzle Fuze",
	'version_major', 1,
	'version_minor', 9,
	'version', 69,
	'lua_revision', 1009413,
	'saved_with_revision', 1010838,
	'code', {
		"Code/TrackedAchievement.lua",
		"Code/AchievementTracker.lua",
		"Code/OSD.lua",
		"Code/CCC.lua",
	},
	'saved', 1645033554,
	'has_options', 'true',
	"TagInterface", true,
})

Code/AchievementTracker.lua
-- See license.md for copyright info
FF.Lib.Debug = false

local function Log(...)
    FF.Funcs.LogMessage(CurrentModDef.title, "AchievementTracker", ...)
end

--shared
SharedModEnv["FF_AT_ShowMainAchievementWindow"] = true

--locals
local AchievementObjects = {}
local InitDone = false

-- hook opening dialogs
local FF_OpenDialog = OpenDialog
function OpenDialog(DialogPath, ...)
    Log("Opening dialog ", DialogPath)
    local Dialog = FF_OpenDialog(DialogPath, ...)
    if DialogPath == "InGameInterface" then
        CreateOSD()
  --  elseif DialogPath == "ColonyControlCenter" then
--        CreateCCC()
    end
    return Dialog
end

--initialize achievement objects
local function Init()
    InitDone = false
    SharedModEnv["FF_AT_ShowOnScreen"] = CurrentModOptions:GetProperty("ShowOnScreen")

    if not MainCity then
        Log("No city!")
        return
    end

    if not AchievementPresets then
        Log("ERROR", "No Achievements!")
        return
    end

    if MainCity.labels.TrackedAchievements then
        MainCity.labels.TrackedAchievements = nil
    end

    for _,Achievement in pairs(AchievementPresets) do

        if Achievement.save_in then
            if (Achievement.save_in == "picard" and g_AvailableDlc.picard) or
                    (Achievement.save_in == "armstrong" and g_AvailableDlc.armstrong) or
                    (Achievement.save_in == "gagarin" and g_AvailableDlc.gagarin) then
            end
        end

        local AchievementObj = PlaceObj("TrackedAchievement")
        AchievementObj.id = _InternalTranslate(Achievement.id)
        AchievementObj.Name = _InternalTranslate(Achievement.display_name)
        AchievementObj.Description = _InternalTranslate(Achievement.how_to)
        AchievementObj.Image = Achievement.image
        AchievementObj.Target = Achievement.target

        AchievementObjects[AchievementObj.id] = AchievementObj

        CreateGameTimeThread(function()
            Sleep(1000) -- wait for PlaceObj to finish
            --boo for hardcoding
            AchievementObjects.AsteroidHopping.Target = 10
            AchievementObjects.USAResearchedEngineering.Target = #Presets.TechPreset.Engineering
            AchievementObjects.Multitasking.Target = 3
            AchievementObjects.SpaceDwarves.Target = 200
            AchievementObjects.Willtheyhold.Target = 100
            AchievementObjects.ScannedAllSectors.Target = 100
            AchievementObjects.DeepScannedAllSectors.Target = 100
            AchievementObjects.MaxedAllTPs.Target = 400
            AchievementObjects.SpaceExplorer.Target = #Presets.TechPreset.ReconAndExpansion -- = 0... data must not be initialized yet, updated later

            --boo for using the wrong scale
            AchievementObjects.IndiaConvertedWasteRock.Type = "Resource"
            AchievementObjects.BrazilConvertedWasteRock.Type = "Resource"
            AchievementObjects.RussiaExtractedAlot.Type = "Resource"
            AchievementObjects.BlueSunProducedFunding.Type = "Resource"
            AchievementObjects.BlueSunExportedAlot.Type = "Resource"
        end)
    end
    InitDone = true
end


--achievement triggers below
function OnMsg.MarkPreciousMetalsExport(city, _)

    if not InitDone then
        return
    end

    --BlueSunExportedAlot
    if GetMissionSponsor().id == "BlueSun" and UIColony.day < 100 then
        AchievementObjects.BlueSunExportedAlot:UpdateValue(city.total_export)
    end
end

function OnMsg.RocketLanded(rocket)

    if not InitDone then
        return
    end

    --AsteroidHopping
    if rocket:IsKindOf("LanderRocketBase") then
        local has_landed_on_asteroid = ObjectIsInEnvironment(rocket, "Asteroid")
        if has_landed_on_asteroid and rocket.asteroids_visited_this_trip <= 10 then
            AchievementObjects.AsteroidHopping:UpdateValue(#rocket.asteroids_visited_this_trip)
        end
    end
    --Landed50Rockets
    if g_RocketsLandedCount <= AchievementPresets.Landed50Rockets.target then
        AchievementObjects.Landed50Rockets:UpdateValue(g_RocketsLandedCount)
    end
end

function OnMsg.TechResearched(_, research, first_time)
    if not first_time or not InitDone then
        return
    end
    --SpaceExplorer
    if AchievementObjects.SpaceExplorer.Target == 0 then
        AchievementObjects.SpaceExplorer.Target = #Presets.TechPreset.ReconAndExpansion
    end
    for field_id, field in pairs(TechFields) do
        if field:HasMember("save_in") and field.save_in == "picard" then
            local researched, total = research:TechCount(field_id, "researched")
            if researched < total then
                if AchievementObjects.SpaceExplorer.Target > 0 then
                    AchievementObjects.SpaceExplorer:UpdateValue(researched)
                    return
                end
            end
        end
    end
    --USAResearchedEngineering
    if sponsor.id == "NASA" and UIColony.day < 100 and research:TechCount("Engineering", "researched") <= #research.tech_field.Engineering then
        AchievementObjects.USAResearchedEngineering:UpdateValue(#research.tech_field.Engineering)
        --EuropeResearchedBreakthroughs
    elseif sponsor.id == "ESA" and UIColony.day < 100 and research:TechCount("Breakthroughs", "researched") <= AchievementPresets.EuropeResearchedBreakthroughs.target then
        AchievementObjects.EuropeResearchedBreakthroughs:UpdateValue(research:TechCount("Breakthroughs", "researched"))
    end
end

function OnMsg.AsteroidRocketLanded(rocket)

    if not InitDone then
        return
    end

    --Multitasking
    if not rocket:IsKindOf("LanderRocketBase") then
        return
    end
    local num_astroids_visiting = 0
    local loaded_maps = GetLoadedMaps()
    for _, map_id in pairs(loaded_maps) do
        local map_data = ActiveMaps[map_id]
        if map_data.Environment == "Asteroid" then
            local city = Cities[map_id]
            local rockets = city.labels.AllRockets or empty_table
            local vehicles = city.labels.Rover or empty_table
            local buildings = city.labels.Building or empty_table
            if 0 < #vehicles or 0 < #rockets or 0 < #buildings then
                num_astroids_visiting = num_astroids_visiting + 1
            end
        end
    end
    if num_astroids_visiting <= 3 then
        AchievementObjects.Multitasking:UpdateValue(num_astroids_visiting)
    end
end

function OnMsg.NewDay(Day)
    if not InitDone then
        return
    end

    if Day == 100 then
        if GetMissionSponsor().id == "BlueSun" then
            AchievementObjects.BlueSunExportedAlot:SetFailed()
            AchievementObjects.BlueSunProducedFunding:SetFailed()
        elseif GetMissionSponsor().id == "CNSA" then
            AchievementObjects.ChinaTaiChiGardens:SetFailed()
            AchievementObjects.ChinaReachedHighPopulation:SetFailed()
        elseif GetMissionSponsor().id == "Japan" then
            AchievementObjects.JapanTrainedSpecialists:SetFailed()
        elseif GetMissionSponsor().id == "ESA" then
            AchievementObjects.EuropeResearchedBreakthroughs:SetFailed()
            AchievementObjects.EuropeResearchedAlot:SetFailed()
        elseif GetMissionSponsor().id == "ISRO" then
            AchievementObjects.IndiaConvertedWasteRock:SetFailed()
            AchievementObjects.IndiaBuiltDomes:SetFailed()
        elseif GetMissionSponsor().id == "Roscosmos" then
            AchievementObjects.RussiaExtractedAlot:SetFailed()
        elseif GetMissionSponsor().id == "NewArk" then
            AchievementObjects.NewArkChurchHappyColonists:SetFailed()
            AchievementObjects.NewArcChurchMartianborns:SetFailed()
        elseif GetMissionSponsor().id == "NASA" then
            AchievementObjects.USAResearchedEngineering:SetFailed()
        elseif GetMissionSponsor().id == "Brazil" then
            AchievementObjects.BrazilConvertedWasteRock:SetFailed()
        end
    end

    --SpaceDwarves
    local underground_city = Cities[UIColony.underground_map_id]
    if not underground_city then
        return
    end
    local number_underground_colonists = #(underground_city.labels.Colonist or "")
    local total_colonists = #(UIColony.city_labels.labels.Colonist or "")
    if number_underground_colonists == total_colonists and 200 > number_underground_colonists then
        AchievementObjects.SpaceDwarves:UpdateValue(number_underground_colonists)
    end
end

function OnMsg.PreventedCaveIn(_)
    if not InitDone then
        return
    end

    --Willtheyhold
    if PreventedCaveIns < 100 then
        AchievementObjects.Willtheyhold:UpdateValue(PreventedCaveIns)
    end
end

function OnMsg.BuildingInit(bld)
    if not InitDone then
        return
    end

    --ChinaTaiChiGardens
    if UIColony.day <= 100 then
        local sponsor_id = GetMissionSponsor().id
        if sponsor_id == "CNSA" and IsKindOf(bld, "TaiChiGarden") then
            local domes_with_garden = {}
            local label = MainCity.labels.TaiChiGarden or empty_table
            for i = 1, #label do
                domes_with_garden[label[i].parent_dome] = true
            end
            if table.count(domes_with_garden) <= AchievementPresets.ChinaTaiChiGardens.target then
                AchievementObjects.ChinaTaiChiGardens:UpdateValue(table.count(domes_with_garden))
            end
        end
    end
end

function OnMsg.TrainingComplete(building, _)
    if not InitDone then
        return
    end

    --JapanTrainedSpecialists
    if UIColony.day <= 100 and building.training_type == "specialization" and GetMissionSponsor().id == "Japan" then
        if TotalTrainedSpecialists <= AchievementPresets.JapanTrainedSpecialists.target then
            AchievementObjects.JapanTrainedSpecialists:UpdateValue(TotalTrainedSpecialists)
        end
    end
end

function OnMsg.FundingChanged(colony, amount)
    if not InitDone then
        return
    end

    --BlueSunProducedFunding
    if GameTime() > 1 and GetMissionSponsor().id == "BlueSun" and UIColony.day <= 100 and 0 < amount then
        if FundingGenerated <= AchievementPresets.BlueSunProducedFunding.target * 1000000 then
            AchievementObjects.BlueSunProducedFunding:UpdateValue(FundingGenerated)
        end
    end
    --GatheredFunding
    if colony.funding <= AchievementPresets.GatheredFunding.target then
        AchievementObjects.GatheredFunding:UpdateValue(colony.funding)
    end
end

function OnMsg.NewHour(_)
    if not InitDone then
        return
    end

    --EuropeResearchedAlot
    local sponsor = GetMissionSponsor().id
    if sponsor == "ESA" and UIColony.day <= 100 and UIColony:GetEstimatedRP() <= AchievementPresets.EuropeResearchedAlot.target then
        AchievementObjects.EuropeResearchedAlot:UpdateValue(UIColony:GetEstimatedRP())
    end
    --NewArkChurchHappyColonists
    if sponsor == "NewArk" and UIColony.day <= 100 then
        local colonists = UIColony:GetCityLabels("Colonist") or empty_table
        local count = 0
        for _, colonist in ipairs(colonists) do
            if colonist.stat_comfort >= 70 * const.Scale.Stat then
                count = count + 1
                if count <= AchievementPresets.NewArkChurchHappyColonists.target then
                    AchievementObjects.NewArkChurchHappyColonists:UpdateValue(count)
                end
            end
        end
    end
    --RussiaHadManyColonists
    if sponsor == "Roscosmos" and CalcChallengeRating() + 100 >= 500 and #(UIColony:GetCityLabels("Colonist") or empty_table) <= AchievementPresets.RussiaHadManyColonists.target then
        AchievementObjects.RussiaHadManyColonists:UpdateValue(#(UIColony:GetCityLabels("Colonist") or empty_table))
    end
    --SpaceYBuiltDrones (this is hardcoded in to Drone.lua and hasn't been updated for B&B)
    if MainCity.labels.Drone then
        if GetMissionSponsor().id == "SpaceY" and UIColony.day < 100 and #MainCity.labels.Drone <= AchievementPresets.SpaceYBuiltDrones.target then
            AchievementObjects.SpaceYBuiltDrones:UpdateValue(#MainCity.labels.Drone or empty_table)
        end
    end
end

function OnMsg.WasteRockConversion(amount, producers)
    if not InitDone then
        return
    end

    local sponsor = GetMissionSponsor().id
    WasteRockConverted = WasteRockConverted + amount
    if producers.PreciousMetals then
        WasteRockConvertedToRareMetals = WasteRockConvertedToRareMetals + amount
    end
    --IndiaConvertedWasteRock
    if sponsor == "ISRO" and UIColony.day <= 100 and WasteRockConverted / const.ResourceScale <= AchievementPresets.IndiaConvertedWasteRock.target then
        AchievementObjects.IndiaConvertedWasteRock:UpdateValue(WasteRockConverted / const.ResourceScale)
        --BrazilConvertedWasteRock
    elseif sponsor == "Brazil" and UIColony.day <= 100 and WasteRockConvertedToRareMetals / const.ResourceScale <= AchievementPresets.BrazilConvertedWasteRock.target then
        AchievementObjects.BrazilConvertedWasteRock:UpdateValue(WasteRockConverted / const.ResourceScale)
    end
end

function OnMsg.TerraformParamChanged()
    if not InitDone then
        return
    end

    local TerraformTotal = 0
    local TerraformingParameters = {
        "Atmosphere",
        "Temperature",
        "Water",
        "Vegetation"
    }

    for _, TerraformingParameter in ipairs(TerraformingParameters) do
        TerraformTotal = TerraformTotal + GetTerraformParamPct(TerraformingParameter)
    end

    AchievementObjects.MaxedAllTPs:UpdateValue(TerraformTotal)
end

local CheckTraitsAchievements = function()
    if not InitDone then
        return
    end

    if GetAchievementFlags("ColonistWithRareTraits") and GetAchievementFlags("HadColonistWith5Perks") and GetAchievementFlags("HadVegans") then
        return
    end
    local vegans_count = 0
    local colonists = UIColony:GetCityLabels("Colonist") or empty_table
    for i = 1, #colonists do
        local c = colonists[i]
        if not GetAchievementFlags("ColonistWithRareTraits") or not GetAchievementFlags("HadColonistWith5Perks") then
            local perks_count = 0
            local rare_traits_count = 0
            for trait_id, _ in pairs(c.traits) do
                if g_RareTraits[trait_id] then
                    rare_traits_count = rare_traits_count + 1
                end
                local trait_data = TraitPresets[trait_id]
                if trait_data and trait_data.group == "Positive" then
                    perks_count = perks_count + 1
                end
            end
            --ColonistWithRareTraits
            if rare_traits_count <= ColonistWithRareTraits_target then
                AchievementObjects.ColonistWithRareTraits:UpdateValue(rare_traits_count)
            end
            --HadColonistWith5Perks
            if perks_count <= HadColonistWith5Perks_target then
                AchievementObjects.HadColonistWith5Perks:UpdateValue(perks_count)
            end
        end
        if c.traits.Vegan then
            vegans_count = vegans_count + 1
        end
    end
    --HadVegans
    if vegans_count <= HadVegans_target then
        AchievementObjects.HadVegans:UpdateValue(vegans_count)
    end
end
function OnMsg.ColonistAddTrait()
    DelayedCall(30000, CheckTraitsAchievements)
end

function OnMsg.SectorScanned()
    if not InitDone then
        return
    end

    if GetAchievementFlags("ScannedAllSectors") and GetAchievementFlags("DeepScannedAllSectors") then
        return
    end
    local sector_status_to_number = {
        unexplored = 0,
        scanned = 1,
        ["deep scanned"] = 2
    }

    local SectorsScanned = 0
    local SectorsDeepScanned = 0
    for x = 1, const.SectorCount do
        for y = 1, const.SectorCount do
            if sector_status_to_number[MainCity.MapSectors[x][y].status] == 1 then
                SectorsScanned = SectorsScanned + 1
            elseif sector_status_to_number[MainCity.MapSectors[x][y].status] == 2 then
                SectorsDeepScanned = SectorsDeepScanned + 1
            end
        end
    end

    local SectorCount = 100
    --ScannedAllSectors
    if SectorsScanned <= SectorCount then
        AchievementObjects.ScannedAllSectors:UpdateValue(SectorsScanned)
    end
    --DeepScannedAllSectors
    if SectorsDeepScanned <= SectorCount then
        AchievementObjects.DeepScannedAllSectors:UpdateValue(SectorsDeepScanned)
    end
end

local CountNonConstructionSitesInLabel = function(city, label)
    local container = (city or UICity).labels[label] or empty_table
    local count = 0
    for i = 1, #container do
        if not IsKindOf(container[i], "ConstructionSite") then
            count = count + 1
        end
    end
    return count
end
function OnMsg.ConstructionComplete(bld)
    if not InitDone then
        return
    end

    if IsKindOf(bld, "RocketLandingSite") then
        return
    end
    local city = bld.city
    --Built1000Buildings
    if g_BuildingsBuilt <= AchievementPresets.Built1000Buildings.target then
        AchievementObjects.Built1000Buildings:UpdateValue(g_BuildingsBuilt)
    end
    --IndiaBuiltDomes
    if IsKindOf(bld, "Dome") and GetMissionSponsor().id == "ISRO" and UIColony.day < 100 and not GetAchievementFlags("IndiaBuiltDomes") and
            CountNonConstructionSitesInLabel(city, "Dome") <= AchievementPresets.IndiaBuiltDomes.target - 1 then
        AchievementObjects.IndiaBuiltDomes:UpdateValue(CountNonConstructionSitesInLabel(city, "Dome"))
    end
    --BuiltSeveralWonders
    if bld.build_category == "Wonders" and not GetAchievementFlags("BuiltSeveralWonders") and
            CountNonConstructionSitesInLabel(city, "Wonders") <= AchievementPresets.BuiltSeveralWonders.target - 1
    then
        AchievementObjects.BuiltSeveralWonders:UpdateValue(CountNonConstructionSitesInLabel(city, "Wonders"))
    end
end

function OnMsg.ResourceExtracted()
    if not InitDone then
        return
    end

    --RussiaExtractedAl1ot
    if GetMissionSponsor().id == "Roscosmos" and UIColony.day < 100 and g_TotalExtractedResources <= RussiaExtractedAlot_target then
        AchievementObjects.RussiaExtractedAlot:UpdateValue(g_TotalExtractedResources)
    end
end

local CheckColonistCountAchievements = function()
    local total_colonists = #(UIColony:GetCityLabels("Colonist") or empty_table)
    --ChinaReachedHighPopulation
    if GetMissionSponsor().id == "CNSA" and UIColony.day < 100 and total_colonists <= ChinaReachedHighPopulation_target then
        AchievementObjects.ChinaReachedHighPopulation:UpdateValue(total_colonists)
    end
    --Reached1000Colonists
    if total_colonists <= Reached1000Colonists_target then
        AchievementObjects.Reached1000Colonists:UpdateValue(total_colonists)
    end
    --Reached250Colonists
    if total_colonists <= Reached250Colonists_target then
        AchievementObjects.Reached250Colonists:UpdateValue(total_colonists)
    end
end
function OnMsg.ColonistBorn(colonist)
    if not InitDone then
        return
    end

    --NewArcChurchMartianborns
    if colonist.traits.Child and colonist.age == 0 then
        if GetMissionSponsor().id == "NewArk" and UIColony.day < 100 and g_TotalChildrenBornWithMating <= AchievementPresets.NewArcChurchMartianborns.target then
            AchievementObjects.NewArcChurchMartianborns:UpdateValue(g_TotalChildrenBornWithMating)
        end
    end
    DelayedCall(1000, CheckColonistCountAchievements)
end
function OnMsg.ColonistArrived()
    DelayedCall(1000, CheckColonistCountAchievements)
end

function OnMsg.ColonistCured(_, bld)
    if not InitDone then
        return
    end

    --CuredColonists
    if bld.total_cured <= AchievementPresets.CuredColonists.target then
        AchievementObjects.CuredColonists:UpdateValue(bld.total_cured)
    end
end

function OnMsg.ColonistJoinsDome(_, dome)
    if not InitDone then
        return
    end

    --Had100ColonistsInDome
    if #(dome.labels.Colonist or empty_table) <= AchievementPresets.Had100ColonistsInDome.target then
        AchievementObjects.Had100ColonistsInDome:UpdateValue(#(dome.labels.Colonist or empty_table))
    end
    --Had50AndroidsInDome
    if #(dome.labels.Android or empty_table) >= AchievementPresets.Had50AndroidsInDome.target then
        AchievementObjects.Had50AndroidsInDome:UpdateValue(#(dome.labels.Android or empty_table))
    end
end

--event handling
function OnMsg.ModsReloaded()
    if MainCity then
        MainCity.labels.TrackedAchievement = nil
        Init()
    end
end
function OnMsg.ApplyModOptions()
    SharedModEnv["FF_AT_ShowMainAchievementWindow"] = CurrentModOptions:GetProperty("ShowMainAchievementWindow")
    CreateOSD()
end
OnMsg.CityStart = Init
OnMsg.LoadGame = Init

items.lua
return {
PlaceObj('ModItemCode', {
	'name', "Achievement",
	'comment', "Class Def",
	'FileName', "Code/TrackedAchievement.lua",
}),
PlaceObj('ModItemCode', {
	'name', "AchievementTracker",
	'comment', "Tracking Code",
	'FileName', "Code/AchievementTracker.lua",
}),
PlaceObj('ModItemCode', {
	'name', "On Screen Display",
	'comment', "OSD for Primary Achievement",
	'FileName', "Code/OSD.lua",
}),
PlaceObj('ModItemCode', {
	'name', "Colony Control Center",
	'comment', "CCC (Command Center) Additions for Achievements",
	'FileName', "Code/CCC.lua",
}),
PlaceObj('ModItemOptionToggle', {
	'name', "ShowFailures",
	'DisplayName', "Show Failure Messages",
	'Help', "Show a message when an achievement is failed",
	'DefaultValue', true,
}),
PlaceObj('ModItemOptionToggle', {
	'name', "ShowMainAchievementWindow",
	'DisplayName', "Show Primary Achievement Window",
	'Help', "Show the main achievement window in the top center of the screen, just below the infopanel.",
	'DefaultValue', true,
}),
PlaceObj('ModItemOptionChoice', {
	'name', "ShowOnScreen",
	'DisplayName', "Primary Achievement",
	'Help', "The primary achievement is always tracked on the top right of the screen",
	'DefaultValue', "Space Communism",
	'ChoiceList', {
		"Building a Better Future",
		"Asteroid Hopping",
		"You can't take the Sky from Me!",
		"Space Explorer",
		"For the Benefit of All",
		"Will they hold?",
		"Europa Universalis",
		"No Pain, No Gain",
		"Waste Not, Want Not",
		"Posthuman",
		"S.P.E.C.I.A.L.",
		"Dream of a Green Mars",
		"Can't Stop the Signal",
		"The Final Frontier",
		"Wubba, lubba, dub, dub!",
		"In the Service of Humankind",
		"The New Wonders of the World",
		"Gagarin's Legacy",
		"Interesting Times",
		"Multiplanet Species",
		"Space Communism",
		"The New Ark",
		"Assisted Self-Improvement",
		"Marsopolis",
		"Do Androids Dream of Electric Sheep?",
		"Aren't they Cute?",
	},
}),
PlaceObj('ModItemOptionNumber', {
	'name', "MessageLength",
	'DisplayName', "Message Length (Real-Time Seconds)",
	'Help', "How long the message will show on the screen in seconds.",
	'DefaultValue', 10,
	'MinValue', 1,
	'MaxValue', 60,
}),
PlaceObj('ModItemOptionNumber', {
	'name', "SolDelay",
	'DisplayName', "Message Delay (Sols)",
	'Help', "Delay in sols before showing another achievement progress message",
	'DefaultValue', 0,
	'MinValue', 0,
	'MaxValue', 7,
}),
PlaceObj('ModItemOptionNumber', {
	'name', "HourDelay",
	'DisplayName', "Message Delay (Hours)",
	'Help', "Delay in hours before showing another achievement progress message",
	'DefaultValue', 1,
	'MinValue', 0,
	'MaxValue', 25,
}),
}